### ISSUE TITLE
Unable to build with prod settings

### ISSUE THREAD

**Issue by kshkrao3** [matheushf/ng2-date-countdown#4]
I am getting below error whenever I execute `ng build --prod`

**Unexpected value 'CountDown in {projectDir}/node_modules/ng2-date-countdown/dist/countdown.d.ts' declared by the module 'AppModule in {projectDir}/src/app/app.module.ts'. Please add a @Pipe/@Directive/@Component annotation.**

**dependencies**
 
```
"dependencies": {
        "@agm/core": "^1.0.0-beta.2",
        "@angular/animations": "^5.0.0",
        "@angular/cdk": "^5.0.0-rc0",
        "@angular/common": "^5.0.0",
        "@angular/compiler": "^5.0.0",
        "@angular/core": "^5.0.0",
        "@angular/forms": "^5.0.0",
        "@angular/http": "^5.0.0",
        "@angular/material": "^5.0.0-rc0",
        "@angular/material-moment-adapter": "^5.0.3",
        "@angular/platform-browser": "^5.0.0",
        "@angular/platform-browser-dynamic": "^5.0.0",
        "@angular/router": "^5.0.0",
        "@types/lodash": "^4.14.65",
        "angular2-moment": "^1.7.1",
        "angularfire2": "^5.0.0-rc.4",
        "classlist.js": "^1.1.20150312",
        "core-js": "^2.4.1",
        "firebase": "^4.6.2",
        "hammerjs": "^2.0.8",
        "intl": "^1.2.5",
        "jquery": "^3.2.1",
        "ng2-cookies": "^1.0.12",
        "ng2-date-countdown": "0.0.5",
        "ng2-slim-loading-bar": "^4.0.0",
        "ng2-validation": "^4.2.0",
        "ngx-pagination": "^3.0.0",
        "ngx-slick": "^0.1.3",
        "razorpay": "^1.6.1",
        "reflect-metadata": "^0.1.12",
        "rxjs": "^5.5.2",
        "sweetalert2": "^7.3.0",
        "web-animations-js": "^2.3.1",
        "zone.js": "^0.8.14"
    },
    "devDependencies": {
        "@angular/cli": "^1.5.4",
        "@angular/compiler-cli": "^5.0.0",
        "@angular/language-service": "^5.0.0",
        "@types/jasmine": "~2.5.53",
        "@types/jasminewd2": "~2.0.2",
        "@types/jquery": "^2.0.47",
        "@types/node": "~6.0.60",
        "codelyzer": "~3.2.0",
        "jasmine-core": "~2.6.2",
        "jasmine-spec-reporter": "~4.1.0",
        "karma": "~1.7.0",
        "karma-chrome-launcher": "~2.1.1",
        "karma-cli": "~1.0.1",
        "karma-coverage-istanbul-reporter": "^1.2.1",
        "karma-jasmine": "~1.1.0",
        "karma-jasmine-html-reporter": "^0.2.2",
        "protractor": "~5.1.2",
        "ts-node": "~3.2.0",
        "tslint": "~5.7.0",
        "typescript": "^2.4.2"
    }
```

**Comment by Albejr** [comment]
`ng build --prod --aot=false`

**Comment by kshkrao3** [comment]
@Albejr But that effects the performance of the application right?

**Comment by matheushf** [comment]
How are you importing it? Especially in app.component
(see last answer from here https://github.com/angular/angular/issues/16502)

And also try doing this:

replace:

```
import 'core-js/es6/reflect';
import 'core-js/es7/reflect';
```
in polyfills.ts with:

`import 'reflect-metadata';`

from https://github.com/angular/angular/issues/15890

**Comment by kshkrao3** [comment]
@matheushf I've tried adding it, unfortunately, the issue was still there. But now I've found a solution. We need to build `aot` compatible version of the library. I just succeeded with the help of **[this post](https://medium.com/@trekhleb/how-to-create-aot-jit-compatible-angular-4-library-with-external-scss-html-templates-9da6e68dac6e)** and yea it took me a while to understand. If you permit me I can contribute to this repository, We just need to modify `package.json`, `tsconfig-build.json` and `gulpfile.js` and build the library again, which adds metadata info for the library. :) The important thing is that we need to exclude `@angular/core` and `@angular/common` if we are targetting angular applications since it will be already included. We can only keep it for `devDependencies` and exclude it via `peerDependencies` as suggested in the post. :) Let me know if I can help.. :)

**Comment by mcblum** [comment]
@kshkrao3 hey how did you end up fixing this particular problem? We're seeing the same thing as you did but I'm not quite sure I understand your fix :) Did you end up just forking it and making the modifications so it would build for AOT?

**Comment by kshkrao3** [comment]
@mcblum Nope. If you follow [this link](https://medium.com/@trekhleb/how-to-create-aot-jit-compatible-angular-4-library-with-external-scss-html-templates-9da6e68dac6e)  you'll come to know the actual problem. As a small attempt, try removing **node_modules** directory located under this library when you have it in your project. Let me know for any issues. Am happy to help.. :)

**Comment by mcblum** [comment]
@kshkrao3 ah yep, that makes sense. I actually did manage to solve this by just copying the component file, making a few changes to the way some variables were defined, adding an `@Output()` for every time the timer updates and including the component directly. Going to try to make a pull request after work.

**Comment by matheushf** [comment]
Oh well, I hope it works

**Comment by kshkrao3** [comment]
@mcblum Am way far from understanding how these pull requests work and how we must commit the changes. I tried but couldn't succeed. Anyways, if you are doing it then it'll be helpful for all. Thank you.. :)

**Comment by tokiodeo** [comment]
my solution was to create the component decorator  **@Input()**  **countdown.d.ts**
import { EventEmitter, Input } from '@angular/core';
import { Component } from '@angular/core';
@Component({
    selector: 'countdown',
    template: `
    <div class="countdown">
      <div *ngFor="let time of display; let i = index" class="measurements {{time}}">
        <p class="measurements-number">
          {{ (showZero && displayNumbers[i] < 10) ? '0' + displayNumbers[i].trim() : displayNumbers[i]}}
        </p>
        <span *ngIf="display[i+1]" class="divider"> {{divider}} </span>
        <p class="measurements-text">
          {{time}}
        </p>
      </div>
    </div>
    <ng-content></ng-content>
  `,
        styles: [`
    .countdown{display:flex;align-items:center;justify-content:center;align-content:center;font-family:"Helvetica Neue", Helvetica, Arial, sans-serif}.countdown .measurements{flex:0.1}.countdown .measurements .divider{float:right}
  `]
  })
export  class CountDown {
    @Input() units: any;
    @Input() end: any;
    @Input() displayString: String;
    @Input() text: any;
    @Input() divider: any;
    @Input() showZero: Boolean;
    @Input() reached: EventEmitter<Date>;
    @Input() display: any;
    @Input() displayNumbers: any;
    @Input() wasReached: Boolean;
     constructor();
    _displayString(): void;
}

-------------------------------------------------------------------------------------------------
I also delete the file decorator  **countdown.js**

`CountDown = tslib_1.__decorate([
    tslib_1.__metadata("design:paramtypes", [])
], CountDown);`

-------------------------------------------------------------------------------------------------

[llink screen shot ng build --prod aot](http://subirimagen.me/uploads/20181021023214.png)

**Comment by tokiodeo** [comment]
You can copy the github component and install it in your project as a separate component of node_modules.

**Comment by franbenz** [comment]
I avoided this issue the same way @jersonjcnt did. I removed the dependency from `package.json` and copied `countdown.html, countdown.scss and countdown.ts` to my own project and referenced it in app.module.ts.

**Comment by matheushf** [comment]
I tried it again with Angular 7, and it worked just fine with --aot and --prod. I'll close the issue until further notices.

**Comment by IRCraziestTaxi** [comment]
Seems to break --prod build in Angular 8.

**Comment by jidorn** [comment]
I face the same issue than @IRCraziestTaxi and @kshkrao3 : --prod is break and go the Error: 
"ERROR in Unexpected value 'CountDown in C: ... node_modules/ng2-date-countdown/dist/countdown.d.ts' declared by the module 'AppModule [...]
 Please add a @Pipe/@Directive/@Component annotation."
Does it will be compatible with Angular 9 ?

**Comment by matheushf** [comment]
Ok guys, now it's serious. It's finally **WORKING** with Prod and AOT. 

I've released it, so just update to `1.1.0` and it should work fine. 

It will be compatible with Angular 9, as it's now with 8.
